runtime/src/bank.rs

```diff
@@ -979,6 +982,9 @@ pub struct Bank {
     /// The number of transactions processed without error
     transaction_count: AtomicU64,
 
+    /// The number of non-vote transactions processed without error
+    non_vote_transaction_count: AtomicU64,
+
```

Comment:
There are `transaction_error_count` and `transaction_entries_count` that are also transient.
Maybe if I just explain the transient nature of this count it would be enough?
Or do you still think the name should be different?

I'm not sure what a good name would be.
`non_vote_transaction_count_transient`?
Seems a bit too long.

Reply:
> There are transaction_error_count and transaction_entries_count that are also transient.

In fairness, these two fields you compared to aren't transient, they are per-bank (slot):
https://github.com/solana-labs/solana/blob/58d66e11133f4cf40dd4a7f6bfeed4b358393fe7/runtime/src/bank.rs#L982-L986
As we see evidenced when we make a new bank:
https://github.com/solana-labs/solana/blob/58d66e11133f4cf40dd4a7f6bfeed4b358393fe7/runtime/src/bank.rs#L1585-L1586

I'll think on this a bit more, but maybe something along the lines of `non_vote_transaction_count_since_restart`. I don't love the use of `restart`, but something like this makes it very clear that this counter resets every time the validator process restarts.

Reply:
Another idea - maybe `non_vote_transaction_count_relative`, and then the comment explains what it is relative to (the snapshot slot)

Reply:
I kind of like `non_vote_transaction_count_since_restart`, though I get why you're uncomfortable about the word restart. I haven't thought of a better word.

Some context that might help the naming: I am planning to add 2 new persistent fields to capture (a) total and (b) non-vote committed transaction for a single bank. Some sort of scheme that could be consistent for the cumulative fields vs the per-bank fields would be nice.

Reply:
I must admit that when I was writing code for `non_vote_transaction_count` I did not fully understand how it interacts with validator restarts and execution.  And was mostly just copying the `transaction_count` processing code.
But, as written right now, `non_vote_transaction_count` is inherited from the parent block:
```rust
            transaction_count: AtomicU64::new(parent.transaction_count()),
            non_vote_transaction_count: AtomicU64::new(parent.non_vote_transaction_count()),
            transaction_error_count: AtomicU64::new(0),
            transaction_entries_count: AtomicU64::new(0),
            transactions_per_entry_max: AtomicU64::new(0),
```
IIUC, `blockstore` content is persistent.  Meaning, as currently written, `non_vote_transaction_count` is preserved across validator restarts, but is not restored when validator recreates ledger from the network from scratch.

I do not know if there is a term that means something like a "slot when the validator created this ledger".
But this is what should be used here, right?
So it could be `non_vote_transaction_count_since_ledger_start` or something like that.

I've renamed to `non_vote_transaction_count_since_restart` for now.
But if you'll come up with a better name, or will decide that this is not the right behavior at all - I'll be glad to change it.

> Some context that might help the naming: I am planning to add 2 new persistent fields to capture (a) total and (b) non-vote committed transaction for a single bank. Some sort of scheme that could be consistent for the cumulative fields vs the per-bank fields would be nice.

To add clarity, it could be nice to split those different kinds of stats into separate structures.
Instead of keeping everything flat in the `Bank`, some additional structure could both make names shorter and make the code more readable.  Something like this:

```rust
/// Stats accumulated across all blocks, from the genesis, block 0.
pub struct BankFromGenesisStats {
    /// The number of transactions processed without error
    transactions: AtomicU64,
}

/// Stats that are counted for a given block.
/// TODO Should it be `BankPerSlotStats` or `BankPerBankStats`?
pub struct BankPerSlotStats {
    /// The number of transaction errors
    transaction_error: AtomicU64,

    /// The number of transaction entries
    transaction_entries: AtomicU64,

    /// The max number of transaction in an entry
    transactions_per_entry_max: AtomicU64,
}

/// Stats accumulated since the ledger creation.  Not shared though the network,
/// but preserved between validator restarts.
pub struct BankSinceLedgerStartStats {
    /// The number of non-vote transactions processed without error since the creation of the
    /// current ledger.  This value is not shared though the network, but is preserved between
    /// validator restarts.
    non_vote_transactions: AtomicU64,
}

pub struct Bank {
    // [...]

    from_genesis: BankFromGenesisStats,

    per_slot: BankPerSlotStats,

    since_ledger_start: BankSinceLedgerStartStats,

    // [...]
}
```

As the structs explicitly say that they hold stats, `_count` suffix becomes redundant.

Reply:
I do like the idea of separate structures to organize the data by lifetime. Let's refactor the existing fields in a separate PR. Related: https://github.com/solana-labs/solana/issues/29400

Reply:
I'll prepare a separate PR, if you think it makes sense.

Should I wait for #29400 or do you want to do this first?
I could also do the refactoring you describe in #29400.
It seems relatively safe, if I just move code around.
Compiler should be able to check that I didn't miss anything.

Reply:
I don't think the refactor you proposed needs to wait on #29400 , so feel free to PR that any time.
If you are interested in working on the bigger Bank refactor, that would be cool! But no pressure. It's not time-sensitive, and I know you have a few other things on your docket.